library/core/src/iter/adapters/filter.rs

```diff
@@ -27,6 +27,41 @@ impl<I, P> Filter<I, P> {
     }
 }
 
+impl<I, P> Filter<I, P>
+where
+    I: Iterator,
+    P: FnMut(&I::Item) -> bool,
+{
+    #[inline]
+    fn next_chunk_dropless<const N: usize>(
+        &mut self,
+    ) -> Result<[I::Item; N], array::IntoIter<I::Item, N>> {
+        let mut array: [MaybeUninit<I::Item>; N] = MaybeUninit::uninit_array();
+        let mut initialized = 0;
+
+        let result = self.iter.try_for_each(|element| {
+            let idx = initialized;
+            initialized = idx + (self.predicate)(&element) as usize;
+
+            // SAFETY: Loop conditions ensure the index is in bounds.
+            unsafe { array.get_unchecked_mut(idx) }.write(element);
```

Comment:
I know this is unrelated to this PR, and is rather about the original code with the leak issue, but I think that when we have a non-leaking, non-unnecessarily-bit-copying-discarded-items, all in all simpler, alternative available:

```suggestion
            if (self.predicate)(&element) {
                unsafe { array.get_unchecked_mut(initialized) }.write(element);
                initialized += 1;
            }
```

then, if we still want to stick to these now needed leak-avoiding hoops and `idx/initialized` split (I assume the objective then is to produce branchless code since in practice it may pan out better?), at the very least it ought to warrant some kind of comment saying so (_e.g._, pointing at benchmarks or codegen comparisons).

  - cc @rossmacarthur, original author of this logic